haskell/private/providers.bzl

```diff
@@ -124,20 +135,38 @@ def get_solibs_for_ghc_linker(hs, build_info):
       build_info: HaskellBinaryInfo provider.
 
     Returns:
-      List of library files suitable for GHC's builtin linker.
+      (library_deps, ld_library_deps)
+      library_deps: List of library files suitable for GHC's builtin linker.
+      ld_library_deps: List of library files that should be available for
+        dynamic loading.
     """
     trans_link_ctx = build_info.transitive_cc_dependencies.dynamic_linking
-    trans_libs = get_mangled_libs(trans_link_ctx.libraries_to_link.to_list())
-    trans_import_libs = set.to_list(build_info.transitive_import_dependencies)
 
-    _library_deps = trans_libs + trans_import_libs
+    unmangled_libs = get_unmangled_libs(trans_link_ctx.libraries_to_link.to_list())
+    mangled_libs = get_mangled_libs(trans_link_ctx.libraries_to_link.to_list())
+    import_libs = set.to_list(build_info.transitive_import_dependencies)
```

Comment:
In this case it might be slightly faster do do both in one manual list comprehension, this way you only do one pass.

Reply:
Hmm, I don't know how to create two lists out of a single list comprehension. But, I've turned it into a single traversal.